### PR TITLE
Updates to the BlsAccount class

### LINKED ISSUE #11: Use BlsProvider and BlsSigner in ethdk

### PR DESCRIPTION
**Pull request by blakecduncan**
Making the following updates to the blsAccount class

- Adding the blsProvider and the blsSigner to the class
- Added in the `resetAccountPrivateKey` function
- Additionally I updated the createAccount function to return the right types for accounts that are created

Resolves: https://github.com/web3well/ethdk/issues/11

### CODE REVIEW

File: ethdk/src/Ethdk.ts [hunk 1]
```diff
@@ -1,28 +1,35 @@
...
   privateKey?: string
   network?: string
 }
 
+interface AccountTypeMap {
+  bls: BlsAccount
+  // Add more types as needed
+}
+
+type AccountTypeToReturnType<T extends keyof AccountTypeMap> = AccountTypeMap[T]
```
Comment (blakecduncan) [inline ethdk/src/Ethdk.ts:14]: I didn't want to have to add type guard check everywhere an account was created which is why I did this. 

Old way to call Bls specific functions on BlsAccounts
```typescript
const blsAccount = createAccount({ accountType: 'bls' })
if (!isBlsAccount(blsAccount)) {
  throw new Error('not bls account');
}
// can now call bls functions
blsAccount.setTrustedAccount(...);
```

After these changes, the returned account has the right type.

```typescript
const blsAccount = createAccount({ accountType: 'bls' });
blsAccount.setTrustedAccount(...)
```

File: ethdk/src/Bls/BlsAccount.ts [hunk 2]
```diff
@@ -97,21 +120,52 @@ export default class BlsAccount implements Account {
...
   }
 
+  /**
+   * Recovers a compromised BLS wallet by assigning it a new private key. This function
+   * must be called from the trusted account that was previously set by the compromised wallet.
+   *
+   * @param compromisedAccountAddress The address of the compromised BLS wallet.
+   * @param recoveryPhrase The recovery phrase associated with the compromised wallet.
+   * @param newPrivateKey The new private key to be assigned to the compromised wallet.
+   * @returns
```
Comment (JohnGuilding) [inline ethdk/src/Bls/BlsAccount.ts]: Unused @returns

File: ethdk/src/Ethdk.ts (hunk 1 above)
Comment (JohnGuilding) [reply ethdk/src/Ethdk.ts:14]: niceee

File: ethdk/src/Ethdk.ts [hunk 3]
```diff
@@ -1,28 +1,35 @@
...
  */
-export async function createAccount({
+export async function createAccount<T extends keyof AccountTypeMap>({
   accountType,
   privateKey,
   network,
-}: AccountConfig): Promise<Account> {
-  if (accountType === BlsAccount.accountType) {
-    return await BlsAccount.createAccount({
+}: AccountConfig & { accountType: T }): Promise<AccountTypeToReturnType<T>> {
```
Comment (JohnGuilding) [inline ethdk/src/Ethdk.ts:26]: Out of interest, what happens if you don't include ` & { accountType: T }`, assume it is type safety but don't quite understand the mechanism

**Review by JohnGuilding — APPROVED** [review]
LGTM, 1 nitpick comment and 1 question

File: ethdk/src/Ethdk.ts (hunk 3 above)
Comment (blakecduncan) [reply ethdk/src/Ethdk.ts:26]: I tried thinking of a way to not include that part cause it's clunky but I added in an empty EoaAccount class and the typing doesn't work without it.

It's basically how we map the value we pass in as the `accountType` param to the keys of the `AccountTypeMap` above.  Then the function knows which account type to use for the return type based off of what is in the mapping above.

Comment (JohnGuilding) [reply ethdk/src/Ethdk.ts:26]: ok yeah that makes sense, thanks

**Pull request merged**